compiler/record-hessian/src/HessianComputer.cpp

```diff
@@ -0,0 +1,195 @@
+  }
+  else
+  {
+    throw std::runtime_error("Unsupported node rank");
+  }
+  length = num_elements / size_in_ch;
+
+  std::vector<float> hessian(size_in_ch * size_in_ch, 0);
+
+  for (int i = 0; i < size_in_ch; ++i)
+  {
+    for (int j = 0; j < size_in_ch; ++j)
+    {
+      float sum = 0;
+      for (int k = 0; k < length; ++k)
+      {
+        sum += buf[i + k * size_in_ch] * buf[j + k * size_in_ch];
+      }
+      hessian[i * size_in_ch + j] = 2 * sum;
+    }
+  }
+
+  HessianVector &vector = _hessian_map[node];
+  vector.update(hessian);
+}
+
+void HessianComputer::recordHessianForConv2D(const luci::CircleNode *node,
+                                             const luci_interpreter::Tensor *input_tensor)
+{
+  assert(input_tensor->shape().rank() == 4);
+  assert(input_tensor->element_type() == DataType::FLOAT32);
+  const auto node_filter = loco::must_cast<luci::CircleConst *>(
+    loco::must_cast<const luci::CircleConv2D *>(node)->filter());
+  const auto node_bias =
+    loco::must_cast<luci::CircleConst *>(loco::must_cast<const luci::CircleConv2D *>(node)->bias());
+  assert(node_filter.dtype() == loco::DataType::FLOAT32);
+  assert(node_bias.dtype() == loco::DataType::FLOAT32);
+  uint32_t size_in_ch =
+    node_filter->size<loco::DataType::FLOAT32>() / node_bias->size<loco::DataType::FLOAT32>();
+
+  uint32_t input_n = input_tensor->shape().dim(0);
+  uint32_t input_h = input_tensor->shape().dim(1);
+  uint32_t input_w = input_tensor->shape().dim(2);
+  uint32_t input_c = input_tensor->shape().dim(3);
+
+  auto circle_conv2d = (luci::CircleConv2D *)node;
+
+  uint32_t stride_h = circle_conv2d->stride()->h();
+  uint32_t stride_w = circle_conv2d->stride()->w();
+  uint32_t dilation_h = circle_conv2d->dilation()->h();
+  uint32_t dilation_w = circle_conv2d->dilation()->w();
+
+  uint32_t kernel_oc = node_filter->dim(0).value();
+  uint32_t kernel_h = node_filter->dim(1).value();
+  uint32_t kernel_w = node_filter->dim(2).value();
+  uint32_t kernel_ic = node_filter->dim(3).value();
+
+  const auto data = input_tensor->data<float>();
+  const auto num_elements = input_tensor->shape().num_elements();
+  std::vector<float> buf(data, data + num_elements);
+
+  unfold(buf, input_n, input_h, input_w, input_c, stride_h, stride_w, dilation_h, dilation_w,
+         kernel_oc, kernel_h, kernel_w, kernel_ic);
+  uint32_t length = buf.size() / size_in_ch;
+
+  std::vector<float> hessian(size_in_ch * size_in_ch, 0);
+  for (int i = 0; i < size_in_ch; ++i)
+  {
+    for (int j = 0; j < size_in_ch; ++j)
+    {
+      float sum = 0;
+      for (int k = 0; k < length; ++k)
+      {
+        sum += buf[i + k * size_in_ch] * buf[j + k * size_in_ch];
+      }
+      hessian[i * size_in_ch + j] = 2 * sum;
+    }
+  }
+
+  HessianVector &vector = _hessian_map[node];
+  vector.update(hessian);
+}
+
+void HessianComputer::recordHessian(const luci::CircleNode *node,
+                                    const luci_interpreter::Tensor *input_tensor)
+{
```

Comment:
I thought I've seen dtype check in this method. have to removed it?
If so, it would be better not and revive conditional checks for this method.